app/gui2/shared/util/data/text.ts

```diff
@@ -0,0 +1,156 @@
+import diff from 'fast-diff'
+import { rangeEncloses, rangeLength, type SourceRange } from '../../yjsModel'
+import { assertEqual } from '../assert'
+
+export type TextEdit = { range: SourceRange; insert: string }
+
+/** Given text and a set of `TextEdit`s, return the result of applying the edits to the text. */
+export function applyTextEdits(oldText: string, textEdits: TextEdit[]) {
+  textEdits.sort((a, b) => a.range[0] - b.range[0])
+  let start = 0
+  let newText = ''
+  for (const textEdit of textEdits) {
+    newText += oldText.slice(start, textEdit.range[0])
+    newText += textEdit.insert
+    start = textEdit.range[1]
+  }
+  newText += oldText.slice(start)
+  return newText
+}
+
+/** Given text before and after a change, return one possible set of `TextEdit`s describing the change. */
+export function textChangeToEdits(before: string, after: string): TextEdit[] {
+  const textEdits: TextEdit[] = []
+  let nextEdit: undefined | TextEdit
+  let pos = 0
+  // Sequences fast-diff emits:
+  // EQUAL, INSERT
+  // EQUAL, DELETE
+  // DELETE, EQUAL
+  // DELETE, INSERT
+  // INSERT, EQUAL
+  for (const [op, text] of diff(before, after)) {
+    switch (op) {
+      case diff.INSERT:
+        if (!nextEdit) nextEdit = { range: [pos, pos], insert: '' }
+        nextEdit.insert = text
+        break
+      case diff.EQUAL:
+        if (nextEdit) {
+          textEdits.push(nextEdit)
+          nextEdit = undefined
+        }
+        pos += text.length
+        break
+      case diff.DELETE: {
+        if (nextEdit) textEdits.push(nextEdit)
+        const endPos = pos + text.length
+        nextEdit = { range: [pos, endPos], insert: '' }
+        pos = endPos
+        break
+      }
+    }
+  }
+  if (nextEdit) textEdits.push(nextEdit)
+  assertEqual(applyTextEdits(before, textEdits), after)
+  return textEdits
+}
+
+/** Translate a `TextEdit` by the specified offset. */
+export function offsetEdit(textEdit: TextEdit, offset: number): TextEdit {
+  return { ...textEdit, range: [textEdit.range[0] + offset, textEdit.range[1] + offset] }
+}
+
+/** Given:
+ *  @param textEdits - A change described by a set of text edits.
+ *  @param spansBefore - A collection of spans in the text before the edit.
+ *  @returns - A sequence of: Each span from `spansBefore` paired with the smallest span of the text after the edit that
+ *  contains all text that was in the original span and has not been deleted. */
+export function applyTextEditsToSpans(textEdits: TextEdit[], spansBefore: SourceRange[]) {
+  // Gather the start and end locations of all spans.
+  const starts = new Array<number>()
+  const ends = new Array<number>()
+  for (const [start, end] of spansBefore) {
+    starts.push(start)
+    ends.push(end)
+  }
+  const numerically = (a: number, b: number) => a - b
+  starts.sort(numerically)
+  ends.sort(numerically)
```

Comment:
Aren't span starts already guaranteed to be sorted? If they **should be**, i'd be fine with an "is sorted" assert here.

If that's the case, then the `starts` array would not even be strictly necessary, once the array shifting is changed to indexed access.